src/OrchardCore.Modules/OrchardCore.Media/Services/MediaFileStorePathProvider.cs

```diff
@@ -0,0 +1,76 @@
+using System;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+using OrchardCore.Environment.Shell;
+using OrchardCore.FileStorage;
+
+namespace OrchardCore.Media.Services
+{
+    public class MediaFileStorePathProvider : IMediaFileStorePathProvider
+    {
+        private readonly string _requestBaseUrl;
+        private readonly string _cdnBaseUrl;
+
+        public MediaFileStorePathProvider(
+            ShellSettings shellSettings,
+            IHttpContextAccessor httpContextAccessor,
+            IOptions<MediaOptions> mediaOptions
+            )
+        {
+            _requestBaseUrl = "/" + IFileStoreExtensions.Combine(null, shellSettings.RequestUrlPrefix, mediaOptions.Value.AssetsRequestPath);
+
+            var originalPathBase = httpContextAccessor
+                .HttpContext?.Features.Get<ShellContextFeature>()?.OriginalPathBase ?? null;
+
+            if (originalPathBase.HasValue)
+            {
+                _requestBaseUrl = IFileStoreExtensions.Combine(null, originalPathBase, _requestBaseUrl);
+            }
+
+            _requestBaseUrl = _requestBaseUrl.TrimEnd('/');
+
+            // Media options configuration ensures any trailing slash is removed.
+            _cdnBaseUrl = mediaOptions.Value.CdnBaseUrl;
+        }
+
+        public string MapPathToPublicUrl(string path)
```

Comment:
I haven't looked at how we could lose this `MapPathToPublicUrl`. Just not sure on how that might work? For all the incoming requests, i.e. `HttpRequest` we don't really need it at all, but for outgoing, i.e. tag helpers, or graphql queries, we still need a way to at the minimum put the assets request path `/media` in front of the file name. 

Open to ideas

Reply:
You are right, this one is actually totally fine as we need to be able to generate a CDN based url, or at least an absolute url. And its implementation should be common to any media store.